docs/discussion/Multi-Tenancy Support.md

```diff
@@ -0,0 +1,285 @@
+4. How does Ratify trust Gatekeeper(mTLS)? Current Ratify follows the first [approach](https://open-policy-agent.github.io/gatekeeper/website/docs/externaldata/#how-the-external-data-provider-trusts-gatekeeper-mtls) recommended by GK. However, it requires Ratify deployed in the same namespace as GK. Therefore, we have to maintain a cluster-wide, well-known CA cert for GK so that Ratify services in all namespaces could trust it.
+
+#### Implementation Detail
+This option requires minimal code changes to Ratify core workflow as Ratify itself doesn't handle the multi-tenancy scenario. However, we have to make a lot of changes to the helm templates/helmfile, making each Ratify service deployed in each namespace. And all Custom Resources should be namespaced instead of clustered.
+
+### Deployment per Cluster
+#### Pros
+1. More resource efficent. Since there is only one Ratify deployment in the cluster, all tenants share the same Ratify pods and dependency resources. If the load traffic increases, admin could easily scale out the Ratify service accordingly.
+2. GK will work in the current way that sends out one ED request to Ratify for each admission review.
+3. Users only need to deploye once when enabling Ratify feature in the cluster. Afterwards, admins can configure CRs in their own namespaces.
+4. Since GK and Ratify can be deployed to the same namespace, we can enable `cert-controller` of GK to set up mTLS automatically.
+
+#### Cons
+1. As current Ratify framework was designed for single tenant use case, we need to make corresponding updates to most components, like verifier/store/certStore/policy.
+2. For distributed cache, take Redis as an example, we should consider if Redis is deployed per namespace or cluster.
+3. For Ratify logs, we can either save logs to the same path or different paths per namespace depending on the decision of isolation level. Same to the filesystem(specifically oras_local_cache).
+4. Unable to configure ResourceQuota and LimitRange for each namespace.
+5. In the external data request from GK to Ratify, it has to add `namespace` besides the existing `images` field though it can be a backward compatible change.
+6. Coarse-grained control over Ratify configurations.
+7. There is limitation to isolate metrics by namespace. But we can add namespace as a label to all metrics for filtering.
+
+#### Implementation Detail
+This option requires minimal updates to the helm templates as Ratify is only deployed to single namespace like before. However, Ratify itself needs to handle the multi-tenancy model. Besides making Custom Resources namespaced, we also need to maintain a mapping from namespace to their own resources, such as verifiers/stores/certStores/policies in Ratify core workflow. Furthermore, depending on how we decide the isolation level of cache/file system/logs, we may partition those data by namespace.
+
+# Conclusion
```

Comment:
Suggestion to add a "open for discussion" section. For example, we still need to discussion when cluster and namespace policy conflict, which one has higher priority.  This list will help us to finalize design of the first iteration.